chaos_monkey/network/tools.py

```diff
@@ -54,26 +64,96 @@ def check_port_tcp(ip, port, timeout=DEFAULT_TIMEOUT, get_banner=False):
             read_ready, _, _ = select.select([sock], [], [], timeout)
             if len(read_ready) > 0:
                 banner = sock.recv(BANNER_READ)
-    except:
+    except socket.error:
         pass
-    
+
     sock.close()
     return True, banner
 
 
-def check_port_udp(ip, port, timeout=DEFAULT_TIMEOUT):
+def check_udp_port(ip, port, timeout=DEFAULT_TIMEOUT):
+    """
+    Checks if a given UDP port is open by checking if it replies to an empty message
+    :param ip:  Target IP
+    :param port: Target port
+    :param timeout: Timeout to wait
+    :return: Tuple, T/F + banner
+    """
     sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
     sock.settimeout(timeout)
-    
+
     data = None
     is_open = False
-    
+
     try:
         sock.sendto("-", (ip, port))
         data, _ = sock.recvfrom(BANNER_READ)
         is_open = True
-    except:
+    except socket.error:
         pass
     sock.close()
 
     return is_open, data
+
+
+def check_tcp_ports(ip, ports, timeout=DEFAULT_TIMEOUT, get_banner=False):
+    """
+    Checks whether any of the given ports are open on a target IP.
+    :param ip:  IP of host to attack
+    :param ports: List of ports to attack. Must not be empty.
+    :param timeout: Amount of time to wait for connection.
+    :param get_banner: T/F if to get first packets from server
+    :return: list of open ports. If get_banner=True, then a matching list of banners.
+    """
+    sockets = [socket.socket(socket.AF_INET, socket.SOCK_STREAM) for _ in range(len(ports))]
+    [s.setblocking(0) for s in sockets]
+    port_attempts = []
+    try:
+        for sock, port in zip(sockets, ports):
+            LOG.debug("Connecting to port %d" % port)
+            err = sock.connect_ex((ip, port))
+            if err == 0:
+                port_attempts.append((port, sock))
+            if err == 10035:  # WSAEWOULDBLOCK is valid, see https://msdn.microsoft.com/en-us/library/windows/desktop/ms740668%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
+                port_attempts.append((port, sock))
+        if len(port_attempts) != 0:
+            num_replies = 0
+            timeout = int(round(timeout))  # clamp to integer, to avoid checking input
+            time_left = timeout
+            write_sockets = []
+            read_sockets = []
+            while num_replies != len(port_attempts):
+                # bad ports show up as err_sockets
+                read_sockets, write_sockets, err_sockets = \
+                    select.select(
+                        [s[1] for s in port_attempts],
+                        [s[1] for s in port_attempts],
+                        [s[1] for s in port_attempts],
+                        time_left)
+                # any read_socket is automatically a writesocket
+                num_replies = len(write_sockets) + len(err_sockets)
+                if num_replies == len(port_attempts) or time_left == 0:
```

Comment:
SLEEP_BETWEEN_POLL=0.5 and timeout is an integer. So for now everything is good.
Change one of them and "time_left == 0" becomes a problem. I think we should change this to "time_left <= 0"